### PR TITLE
Making Dynamical Systems Tests Reparametrizeable

### PR DESCRIPTION
**Pull request by azane**
This PR should not change any functioning of any test, but rather:

1. Exposes certain test arguments in `pytest.mark.parametrize` so that external repositories can reparametrize those tests with alternative types and values.
2. Restructures test fixtures so that external repositories can inherit and "mixin" a subset of the fixture functionality.

Remaining TODOs:

- [x] parametrize solver instances in `test_validate_dynamics`
- [x] parametrize solver instances in `test_handler_composition`
- [x] remove loose allclose tolerance change in lieu of increasing default precision elsewhere.
- [x] remove modularization of gather in favor of just duplicating boilerplate gather overload in external repositories

Blocking [#8 on chirho_diffeqpy](https://github.com/BasisResearch/chirho_diffeqpy/issues/8)

### CODE REVIEW

File: chirho/indexed/internals.py [hunk 1]
```diff
@@ -37,6 +38,21 @@ def _gather_number(
     )
 
 
+@functools.singledispatch
+def index_select_from_array_like(arr, dim: int, indices: List[int]):
```
Comment (azane) [inline chirho/indexed/internals.py:42]: Don't add this here at this level, instead copy/paste the beginning part of gather and change the end that needs to instead use `np.take`

Comment (azane) [reply chirho/indexed/internals.py]: done

File: tests/dynamical/dynamical_fixtures.py [hunk 2]
```diff
@@ -1,39 +1,40 @@
-from typing import TypeVar
+from typing import Mapping, TypeVar
 
 import pyro
 import torch
 from pyro.distributions import Normal, Uniform, constraints
+from torch import Tensor as Tnsr
```
Comment (SamWitty) [inline tests/dynamical/dynamical_fixtures.py:6]: :(

File: tests/dynamical/test_handler_composition.py [hunk 3]
```diff
@@ -59,12 +60,13 @@
 reparam = pyro.poutine.reparam(config=reparam_config)
 
 
-def counterf_model():
-    model = UnifiedFixtureDynamicsReparam(beta=0.5, gamma=0.7)
+def counterf_model(solver, model):
+    # model = UnifiedFixtureDynamicsReparam(beta=0.5, gamma=0.7)
```
Comment (SamWitty) [inline tests/dynamical/test_handler_composition.py:64]: Please remove

File: tests/dynamical/test_handler_composition.py [hunk 4]
```diff
@@ -59,12 +60,13 @@
...
     obs = condition(data=flight_landing_data)(model.observation)
     vec_obs3 = StaticBatchObservation(times=flight_landing_times, observation=obs)
     with vec_obs3:
-        with TorchDiffEq():
+        # with TorchDiffEq():
```
Comment (SamWitty) [inline tests/dynamical/test_handler_composition.py:68]: Please remove

File: tests/dynamical/test_solver.py [hunk 5]
```diff
@@ -25,17 +25,23 @@ def test_no_backend_error():
         simulate(sir, init_state, start_time, end_time)
 
 
-def test_backend_arg():
-    sir = bayes_sir_model()
-    with TorchDiffEq():
-        result = simulate(sir, init_state, start_time, end_time)
+@pytest.mark.parametrize("backend", [TorchDiffEq])
```
Comment (SamWitty) [inline tests/dynamical/test_solver.py:28]: please change the name of the test and the argument name to "solver"

File: tests/dynamical/test_solver.py [hunk 6]
```diff
@@ -25,17 +25,23 @@ def test_no_backend_error():
...
+@pytest.mark.parametrize("dynamics", [bayes_sir_model()])
+def test_backend_arg(backend, dynamics):
+    with backend():
+        result = simulate(dynamics, init_state, start_time, end_time)
     assert result is not None
 
 
-def test_torchdiffeq_broadcasting():
+@pytest.mark.parametrize("backend", [TorchDiffEq])
```
Comment (SamWitty) [inline tests/dynamical/test_solver.py:36]: See above comment

File: tests/dynamical/dynamical_fixtures.py (hunk 2 above)
Comment (SamWitty) [reply tests/dynamical/dynamical_fixtures.py:6]: please remove

File: tests/dynamical/dynamical_fixtures.py [hunk 7]
```diff
@@ -47,9 +48,47 @@ def observation(self, X: State[torch.Tensor]):
         self._unit_measurement_error("R_obs", X["R"])
 
 
-def bayes_sir_model():
+class SIRReparamObservationMixin(SIRObservationMixin):
+    def observation(self, X: State[torch.Tensor]):
+        # super().observation(X)
```
Comment (SamWitty) [inline tests/dynamical/dynamical_fixtures.py:53]: Please remove

**Review by SamWitty — APPROVED** [review]

**Pull request merged**